scripts/src/collate_fastq_reads.R

```diff
@@ -96,136 +39,158 @@ validate_detected_flowcells= function(detected_flowcells, expected_flowcells) {
     
   } else {
-    print('Flowcell_name and/or flowcell_lane were not detected in raw_counts_uncollapsed.')
-    print('Proceeding without filtering flowcells ...')
+    print('Flowcell_name and/or flowcell_lane are not detected in raw_counts_uncollapsed.')
+    print('Proceeding without filtering flowcells.')
   }
-  
-  # Create sequence map ----
-  sequencing_map= sample_meta %>% dplyr::distinct(pick(all_of(c(sequencing_index_cols, id_cols))))
-  
-  # Validation: Check that mapping is one to one ----
-  check_mapping= sequencing_map %>% dplyr::group_by(pick(all_of(sequencing_index_cols))) %>%
-    dplyr::filter(dplyr::n() > 1) %>% dplyr::ungroup()
-  if(nrow(check_mapping) > 0) {
-    print('The following sequening locations map to multiple conditions.')
-    print(check_mapping)
-    stop('The sequencing index columns do not map 1 to 1 to the ID columns.')
+
+  # Validation: Check that sequencing_index_cols exist in uncollapsed_raw_counts ----
+  if(!validate_columns_exist(sequencing_index_cols, uncollapsed_raw_counts)) {
+    stop('Some sequencing_index_cols are NOT present in the uncollapsed_raw_counts')
   }
   
-  # Create raw counts file ----
-  # Filter for the expected flowcells and summed up the reads over the ID cols.
-  print('Summing up reads ...')
-  raw_counts= uncollapsed_raw_counts %>%
-    dplyr::inner_join(sequencing_map, by= sequencing_index_cols, relationship= 'many-to-one') %>%
-    dplyr::group_by(pick(all_of(c(id_cols, barcode_col)))) %>%
-    dplyr::summarize(n= sum(n)) %>% dplyr::ungroup()
+  # Create summed_reads file ----
+  print('Summing up reads.')
+  # Performing inner join with data.table instead of dplyr
+  summed_reads= data.table::merge.data.table(uncollapsed_raw_counts, sequencing_map, by= sequencing_index_cols)
+  # Code below is checking if a barcode is in the list of known barcodes.
+  # If the barcode is not in the list of known barcodes and its counts is below the low_abundance_threshold, 
+  # then the barcode is replaced with the string "unknown_low_abundance_barcode".
+  # Function := performs the mutate in place without copying the dataframe.
+  # Functions fifelse and %chin% are just faster data.table versions of ifelse and %in%.
+  summed_reads[, c(barcode_col) := data.table::fifelse(get(barcode_col) %chin% unique(known_barcodes) |
+                                                         n >= low_abundance_threshold,
+                                                       get(barcode_col), 'unknown_low_abundance_barcode')]
+
+  # Use data.table to group by id_cols and barcode_col and sum up reads across flowcells.
+  summed_reads= summed_reads[, .(n= sum(n)), by= c(id_cols, barcode_col)]
   
   # Calculate index purity ----
-  index_purity= sum(raw_counts$n) / sum(uncollapsed_raw_counts$n)
-  print(paste0('Index purity: ', round(index_purity, 4)))
+  # This is only accurate if the Nori input file is small enough to fit into a chunk.
```

Comment:
if you return the numerator and denominator of the index purity in each chunk instead of the calculating  the index purity in each chunk separately, then the index purity across chunks could be calculated outside this function instead.